File: bench/diff_string.c
// SPDX-License-Identifier: GPL-2.0
// Tier-2.5 differential oracle: C original vs Rust translation, string.
// Reference extracted from lib/string.c (v7.1); kept byte-identical for
// the non-arch-overridden subset (rule 0026 — riscv64 KASAN-off
// overrides memset/memcpy/memmove/strcmp/strlen/strncmp/strnlen/
// strchr/strrchr via arch/riscv/lib/*.S, not exercised here).
#include <ctype.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>

typedef unsigned long ul;
#define E2BIG 7

static int my_strncasecmp(const char *s1, const char *s2, size_t len)
{
	unsigned char c1, c2;
	if (!len) return 0;
	do {
		c1 = *s1++;
		c2 = *s2++;
		if (!c1 || !c2) break;
		if (c1 == c2) continue;
		c1 = tolower(c1);
		c2 = tolower(c2);
		if (c1 != c2) break;
	} while (--len);
	return (int)c1 - (int)c2;
}

static int my_strcasecmp(const char *s1, const char *s2)
{
	int c1, c2;
	do {
		c1 = tolower((unsigned char)*s1++);
		c2 = tolower((unsigned char)*s2++);
	} while (c1 == c2 && c1 != 0);
	return c1 - c2;
}

static char *my_strncpy(char *dest, const char *src, size_t count)
{
	char *tmp = dest;
	while (count) {
		if ((*tmp = *src) != 0) src++;
		tmp++; count--;
	}
	return dest;
}

// sized_strscpy: byte-replicates the real riscv64 word-at-a-time
// algorithm (lib/string.c + arch/riscv/include/asm/word-at-a-time.h).
// The page-boundary-avoidance branch
// (CONFIG_HAVE_EFFICIENT_UNALIGNED_ACCESS, set on riscv64) is skipped
// here deliberately: on a host heap buffer there is no unmapped-next-
// page hazard to guard against, and skipping it does not change
// dest/return-value outputs (which is all this oracle checks) — it
// only changes how conservatively `max` is capped before the loop.
#define ONE_BITS (~0ul / 0xff)
#define HIGH_BITS (ONE_BITS * 0x80)
static inline ul has_zero(ul val, ul *bits)
{
	ul mask = ((val - ONE_BITS) & ~val) & HIGH_BITS;
	*bits = mask;
	return mask;
}
static inline ul create_zero_mask(ul bits)
{
	bits = (bits - 1) & ~bits;
	return bits >> 7;
}
static inline int my_fls64(uint64_t x)
{
	if (x == 0) return 0;
	return 64 - __builtin_clzll(x);
}
static inline ul find_zero(ul mask) { return my_fls64(mask) >> 3; }

static ssize_t my_sized_strscpy(char *dest, const char *src, size_t count)
{
	size_t max = count;
	long res = 0;

	if (count == 0 || count > INT32_MAX)
		return -E2BIG;

	while (max >= sizeof(ul)) {
		ul bits;
		ul c;
		memcpy(&c, src + res, sizeof(ul));
		if (has_zero(c, &bits)) {
			ul data = create_zero_mask(bits);
			ul bytemask = data;
			ul masked = c & bytemask;
			memcpy(dest + res, &masked, sizeof(ul));
			return res + find_zero(data);
		}
		count -= sizeof(ul);
		if (!count) {
			c &= (~0ul >> 8);
			memcpy(dest + res, &c, sizeof(ul));
			return -E2BIG;
		}
		memcpy(dest + res, &c, sizeof(ul));
		res += sizeof(ul);
		max -= sizeof(ul);
	}

	while (count > 1) {
		char c = src[res];
		dest[res] = c;
		if (!c) return res;
		res++;
		count--;
	}

	dest[res] = '\0';
	return src[res] ? -E2BIG : res;
}

static char *my_strncat(char *dest, const char *src, size_t count)
{
	char *tmp = dest;
	if (count) {
		while (*dest) dest++;
		while ((*dest++ = *src++) != 0) {
			if (--count == 0) { *dest = '\0'; break; }
		}
	}
	return tmp;
}

static size_t my_strlcat(char *dest, const char *src, size_t count)
{
	size_t dsize = strlen(dest);
	size_t len = strlen(src);
	size_t res = dsize + len;
	assert(dsize < count);
	dest += dsize;
	count -= dsize;
	if (len >= count) len = count - 1;
	memcpy(dest, src, len);
	dest[len] = 0;
	return res;
}

static char *my_strchrnul(const char *s, int c)
{
	while (*s && *s != (char)c) s++;
	return (char *)s;
}

static char *my_strnchrnul(const char *s, size_t count, int c)
{
	while (count-- && *s && *s != (char)c) s++;
	return (char *)s;
}

static char *my_strnchr(const char *s, size_t count, int c)
{
	while (count--) {
		if (*s == (char)c) return (char *)s;
		if (*s++ == '\0') break;
	}
	return NULL;
}

static size_t my_strspn(const char *s, const char *accept)
{
	const char *p;
	for (p = s; *p != '\0'; ++p)
		if (!strchr(accept, *p)) break;
	return p - s;
}

static size_t my_strcspn(const char *s, const char *reject)
{
	const char *p;
	for (p = s; *p != '\0'; ++p)
		if (strchr(reject, *p)) break;
	return p - s;
}

static char *my_strpbrk(const char *cs, const char *ct)
{
	const char *sc;
	for (sc = cs; *sc != '\0'; ++sc)
		if (strchr(ct, *sc)) return (char *)sc;
	return NULL;
}

static char *my_strsep(char **s, const char *ct)
{
	char *sbegin = *s;
	char *end;
	if (sbegin == NULL) return NULL;
	end = my_strpbrk(sbegin, ct);
	if (end) *end++ = '\0';
	*s = end;
	return sbegin;
}

static int my_memcmp(const void *cs, const void *ct, size_t count)
{
	if (count >= sizeof(ul)) {
		const ul *u1 = cs;
		const ul *u2 = ct;
		do {
			if (*u1 != *u2) break;
			u1++; u2++;
			count -= sizeof(ul);
		} while (count >= sizeof(ul));
		cs = u1; ct = u2;
	}
	const unsigned char *su1, *su2;
	int res = 0;
	for (su1 = cs, su2 = ct; 0 < count; ++su1, ++su2, count--)
		if ((res = *su1 - *su2) != 0) break;
	return res;
}

static void *my_memscan(void *addr, int c, size_t size)
{
	unsigned char *p = addr;
	while (size) {
		if (*p == (unsigned char)c) return (void *)p;
		p++; size--;
	}
	return (void *)p;
}

static char *my_strstr(const char *s1, const char *s2)
{
	size_t l1, l2;
	l2 = strlen(s2);
	if (!l2) return (char *)s1;
	l1 = strlen(s1);
	while (l1 >= l2) {
		l1--;
		if (!my_memcmp(s1, s2, l2)) return (char *)s1;
		s1++;
	}
	return NULL;
}

static char *my_strnstr(const char *s1, const char *s2, size_t len)
{
	size_t l2 = strlen(s2);
	if (!l2) return (char *)s1;
	while (len >= l2) {
		len--;
		if (!my_memcmp(s1, s2, l2)) return (char *)s1;
		s1++;
	}
	return NULL;
}

static void *my_memchr(const void *s, int c, size_t n)
{
	const unsigned char *p = s;
	while (n-- != 0)
		if ((unsigned char)c == *p++) return (void *)(p - 1);
	return NULL;
}

static void *check_bytes8(const uint8_t *start, uint8_t value, unsigned int bytes)
{
	while (bytes) {
		if (*start != value) return (void *)start;
		start++; bytes--;
	}
	return NULL;
}

static void *my_memchr_inv(const void *start, int c, size_t bytes)
{
	uint8_t value = c;
	uint64_t value64;
	unsigned int words, prefix;

	if (bytes <= 16) return check_bytes8(start, value, bytes);

	value64 = value;
	value64 *= 0x0101010101010101ULL;

	prefix = (unsigned long)start % 8;
	if (prefix) {
		uint8_t *r;
		prefix = 8 - prefix;
		r = check_bytes8(start, value, prefix);
		if (r) return r;
		start += prefix;
		bytes -= prefix;
	}

	words = bytes / 8;
	while (words) {
		if (*(uint64_t *)start != value64) return check_bytes8(start, value, 8);
		start += 8;
		words--;
	}
	return check_bytes8(start, value, bytes % 8);
}

// Explicit LCG (same constants as bench/diff_base64.c).
static uint64_t lcg_state;
static uint32_t lcg_next(void)
{
	lcg_state = lcg_state * 6364136223846793005ULL + 1442695040888963407ULL;
	return (uint32_t)(lcg_state >> 32);
}

static const char ALPHABET[] = "abcXYZ .";
static void gen_str(char *buf, int maxlen)
{
	int len = 1 + lcg_next() % (maxlen - 1);
	for (int i = 0; i < len; i++) buf[i] = ALPHABET[lcg_next() % (sizeof(ALPHABET) - 1)];
	buf[len] = 0;
}
static void gen_bytes(unsigned char *buf, int len, int biased)
{
	for (int i = 0; i < len; i++)
		buf[i] = biased ? (lcg_next() % 3) : (lcg_next() & 0xff);
}

#define BUFLEN 64

int main(int argc, char **argv)
{
	long n = argc > 1 ? atol(argv[1]) : 5000;
	lcg_state = argc > 2 ? (uint64_t)atol(argv[2]) : 271828;

	char a[BUFLEN], b[BUFLEN];

	for (long i = 0; i < n; i++) {
		memset(a, 0, sizeof(a));
		memset(b, 0, sizeof(b));
		gen_str(a, BUFLEN - 1);
		gen_str(b, BUFLEN - 1);
		size_t len = lcg_next() % (BUFLEN - 1) + 1;

		printf("strncasecmp,%s,%s,%zu,%d\n", a, b, len, my_strncasecmp(a, b, len));
		printf("strcasecmp,%s,%s,%d\n", a, b, my_strcasecmp(a, b));

		char dst1[BUFLEN] = {0};
		my_strncpy(dst1, a, BUFLEN - 1);
		printf("strncpy,%s,[%.*s]\n", a, BUFLEN - 1, dst1);

		char dst2[BUFLEN];
		memset(dst2, 0x55, sizeof(dst2));
		size_t cnt = lcg_next() % (BUFLEN - 1) + 1;
		ssize_t r = my_sized_strscpy(dst2, a, cnt);
		printf("strscpy,%s,%zu,%zd,[%.*s]\n", a, cnt, r, BUFLEN, dst2);

		char dst3[BUFLEN];
		strcpy(dst3, a);
		my_strncat(dst3, b, lcg_next() % 16);
		printf("strncat,%s,%s,[%s]\n", a, b, dst3);

		char dst4[BUFLEN];
		strcpy(dst4, a);
		size_t lr = my_strlcat(dst4, b, BUFLEN);
		printf("strlcat,%s,%s,%zu,[%s]\n", a, b, lr, dst4);

		int ch = ALPHABET[lcg_next() % (sizeof(ALPHABET) - 1)];
		char *r1 = my_strchrnul(a, ch);
		printf("strchrnul,%s,%c,%ld\n", a, ch, r1 - a);
		char *r2 = my_strnchrnul(a, len, ch);
		printf("strnchrnul,%s,%zu,%c,%ld\n", a, len, ch, r2 - a);
		char *r3 = my_strnchr(a, len, ch);
		printf("strnchr,%s,%zu,%c,%ld\n", a, len, ch, r3 ? r3 - a : -1);

		printf("strspn,%s,%s,%zu\n", a, b, my_strspn(a, b));
		printf("strcspn,%s,%s,%zu\n", a, b, my_strcspn(a, b));
		char *r4 = my_strpbrk(a, b);
		printf("strpbrk,%s,%s,%ld\n", a, b, r4 ? r4 - a : -1);

		char work[BUFLEN];
		strcpy(work, a);
		char *sp = work;
		printf("strsep,%s,%s", a, b);
		for (int k = 0; k < 5 && sp; k++) {
			char *tok = my_strsep(&sp, b);
			printf(",[%s]", tok);
		}
		printf("\n");

		unsigned char m1[BUFLEN], m2[BUFLEN];
		gen_bytes(m1, BUFLEN, 1);
		memcpy(m2, m1, BUFLEN);
		size_t mlen = lcg_next() % BUFLEN;
		if (lcg_next() % 4 == 0 && mlen > 0)
			m2[lcg_next() % mlen] ^= 0xff;
		printf("memcmp,%zu,%d\n", mlen, my_memcmp(m1, m2, mlen));

		unsigned char sbuf[BUFLEN];
		gen_bytes(sbuf, BUFLEN, 1);
		int target = lcg_next() % 3;
		void *sr = my_memscan(sbuf, target, BUFLEN);
		printf("memscan,%d,%ld\n", target, (unsigned char *)sr - sbuf);

		char *sr2 = my_strstr(a, b);
		printf("strstr,%s,%s,%ld\n", a, b, sr2 ? sr2 - a : -1);
		size_t slen = lcg_next() % (BUFLEN - 1);
		char *sr3 = my_strnstr(a, b, slen);
		printf("strnstr,%s,%s,%zu,%ld\n", a, b, slen, sr3 ? sr3 - a : -1);

		void *cr = my_memchr(m1, target, mlen);
		printf("memchr,%d,%zu,%ld\n", target, mlen, cr ? (unsigned char *)cr - m1 : -1);

		unsigned char invbuf[BUFLEN];
		int fillval = lcg_next() % 3;
		for (int k = 0; k < BUFLEN; k++) invbuf[k] = fillval;
		if (lcg_next() % 3 != 0) {
			int pos = lcg_next() % BUFLEN;
			invbuf[pos] = (fillval + 1) % 3;
		}
		size_t ilen = lcg_next() % BUFLEN;
		void *ir = my_memchr_inv(invbuf, fillval, ilen);
		printf("memchr_inv,%d,%zu,%ld\n", fillval, ilen, ir ? (unsigned char *)ir - invbuf : -1);
	}
	return 0;
}

File: bench/diff_string.rs
// SPDX-License-Identifier: GPL-2.0
//! Tier-2.5 differential oracle: Rust translation side, string.
//! Faithful copy of lib/string_rs.rs's algorithms (kernel-crate
//! bindings/export stripped, libc calls replaced with equivalent Rust)
//! — same protocol/LCG as diff_string.c.

use std::ffi::CStr;
use std::os::raw::{c_char, c_int, c_void};

fn tolower(c: u8) -> u8 {
    if c.is_ascii_uppercase() {
        c - b'A' + b'a'
    } else {
        c
    }
}

fn strncasecmp(s1: &[u8], s2: &[u8], len: usize) -> i32 {
    if len == 0 {
        return 0;
    }
    let mut i1 = 0usize;
    let mut i2 = 0usize;
    let mut remaining = len;
    let mut c1: u8;
    let mut c2: u8;
    loop {
        c1 = s1[i1];
        i1 += 1;
        c2 = s2[i2];
        i2 += 1;
        if c1 == 0 || c2 == 0 {
            break;
        }
        if c1 == c2 {
            remaining -= 1;
            if remaining == 0 {
                break;
            }
            continue;
        }
        c1 = tolower(c1);
        c2 = tolower(c2);
        if c1 != c2 {
            break;
        }
        remaining -= 1;
        if remaining == 0 {
            break;
        }
    }
    c1 as i32 - c2 as i32
}

fn strcasecmp(s1: &[u8], s2: &[u8]) -> i32 {
    let mut i1 = 0usize;
    let mut i2 = 0usize;
    let mut c1: i32;
    let mut c2: i32;
    loop {
        c1 = tolower(s1[i1]) as i32;
        i1 += 1;
        c2 = tolower(s2[i2]) as i32;
        i2 += 1;
        if !(c1 == c2 && c1 != 0) {
            break;
        }
    }
    c1 - c2
}

fn strncpy_(dest: &mut [u8], src: &[u8], mut count: usize) {
    let mut ti = 0usize;
    let mut si = 0usize;
    while count != 0 {
        let c = src[si];
        dest[ti] = c;
        if c != 0 {
            si += 1;
        }
        ti += 1;
        count -= 1;
    }
}

const ALLBUTLAST_BYTE_MASK: usize = !0usize >> 8;

fn has_zero(val: usize, one_bits: usize, high_bits: usize) -> usize {
    ((val.wrapping_sub(one_bits)) & !val) & high_bits
}
fn create_zero_mask(bits: usize) -> usize {
    let bits = bits.wrapping_sub(1) & !bits;
    bits >> 7
}
fn fls64(x: u64) -> i32 {
    if x == 0 {
        0
    } else {
        (63 - x.leading_zeros() as i32) + 1
    }
}
fn find_zero(mask: usize) -> usize {
    (fls64(mask as u64) >> 3) as usize
}

// See the matching comment in diff_string.c: the page-boundary-
// avoidance branch is intentionally skipped (no hazard on a host
// buffer, and it doesn't change dest/return-value outputs).
fn sized_strscpy(dest: &mut [u8], src: &[u8], count: usize) -> isize {
    const ONE_BITS: usize = usize::MAX / 0xff;
    const HIGH_BITS: usize = ONE_BITS * 0x80;

    let mut max = count;
    let mut res: isize = 0;

    if count == 0 || count > i32::MAX as usize {
        return -7; // -E2BIG
    }

    let mut count = count;

    while max >= core::mem::size_of::<usize>() {
        let c = usize::from_ne_bytes(
            src[res as usize..res as usize + core::mem::size_of::<usize>()]
                .try_into()
                .unwrap(),
        );
        let mask = has_zero(c, ONE_BITS, HIGH_BITS);
        if mask != 0 {
            let data = create_zero_mask(mask);
            let bytemask = data;
            let masked = c & bytemask;
            dest[res as usize..res as usize + core::mem::size_of::<usize>()]
                .copy_from_slice(&masked.to_ne_bytes());
            return res + find_zero(data) as isize;
        }
        count -= core::mem::size_of::<usize>();
        if count == 0 {
            let c = c & ALLBUTLAST_BYTE_MASK;
            dest[res as usize..res as usize + core::mem::size_of::<usize>()]
                .copy_from_slice(&c.to_ne_bytes());
            return -7;
        }
        dest[res as usize..res as usize + core::mem::size_of::<usize>()]
            .copy_from_slice(&c.to_ne_bytes());
        res += core::mem::size_of::<usize>() as isize;
        max -= core::mem::size_of::<usize>();
    }

    while count > 1 {
        let c = src[res as usize];
        dest[res as usize] = c;
        if c == 0 {
            return res;
        }
        res += 1;
        count -= 1;
    }

    dest[res as usize] = 0;
    if src[res as usize] != 0 {
        -7
    } else {
        res
    }
}

fn strncat_(dest: &mut Vec<u8>, src: &[u8], mut count: usize) {
    if count != 0 {
        let mut si = 0usize;
        loop {
            let c = src[si];
            si += 1;
            dest.pop(); // remove trailing NUL before appending
            dest.push(c);
            dest.push(0);
            if c == 0 {
                break;
            }
            count -= 1;
            if count == 0 {
                break;
            }
        }
    }
}

fn strlcat_(dest: &mut Vec<u8>, src: &[u8], count: usize) -> usize {
    let dsize = dest.len() - 1; // dest is NUL-terminated Vec (excl. trailing 0 in "len")
    let len = src.len();
    let res = dsize + len;
    assert!(dsize < count, "BUG_ON hit: dsize >= count");
    let mut remaining = count - dsize;
    let mut len = len;
    if len >= remaining {
        len = remaining - 1;
    }
    dest.truncate(dsize);
    dest.extend_from_slice(&src[..len]);
    dest.push(0);
    let _ = &mut remaining;
    res
}

fn strchrnul(s: &[u8], c: u8) -> usize {
    let mut i = 0;
    while s[i] != 0 && s[i] != c {
        i += 1;
    }
    i
}

fn strnchrnul(s: &[u8], mut count: usize, c: u8) -> usize {
    let mut i = 0;
    while count != 0 {
        count -= 1;
        if s[i] == 0 || s[i] == c {
            break;
        }
        i += 1;
    }
    i
}

fn strnchr(s: &[u8], mut count: usize, c: u8) -> isize {
    let mut i = 0isize;
    while count != 0 {
        count -= 1;
        if s[i as usize] == c {
            return i;
        }
        if s[i as usize] == 0 {
            break;
        }
        i += 1;
    }
    -1
}

fn strspn(s: &[u8], accept: &[u8]) -> usize {
    let mut i = 0;
    while s[i] != 0 {
        if !accept.contains(&s[i]) {
            break;
        }
        i += 1;
    }
    i
}

fn strcspn(s: &[u8], reject: &[u8]) -> usize {
    let mut i = 0;
    while s[i] != 0 {
        if reject.contains(&s[i]) {
            break;
        }
        i += 1;
    }
    i
}

fn strpbrk(cs: &[u8], ct: &[u8]) -> isize {
    let mut i = 0isize;
    while cs[i as usize] != 0 {
        if ct.contains(&cs[i as usize]) {
            return i;
        }
        i += 1;
    }
    -1
}

fn strsep<'a>(s: &mut Option<&'a [u8]>, ct: &[u8]) -> Option<&'a [u8]> {
    let sbegin = (*s)?;
    let end = strpbrk(sbegin, ct);
    if end >= 0 {
        let tok = &sbegin[..end as usize];
        *s = Some(&sbegin[end as usize + 1..]);
        Some(tok)
    } else {
        *s = None;
        Some(sbegin)
    }
}

fn memcmp_(cs: &[u8], ct: &[u8], count: usize) -> i32 {
    let mut count = count;
    let mut i = 0usize;

    if count >= core::mem::size_of::<usize>() {
        loop {
            let u1 = usize::from_ne_bytes(cs[i..i + 8].try_into().unwrap());
            let u2 = usize::from_ne_bytes(ct[i..i + 8].try_into().unwrap());
            if u1 != u2 {
                break;
            }
            i += 8;
            count -= 8;
            if count < 8 {
                break;
            }
        }
    }

    let mut res: i32 = 0;
    let mut j = i;
    let mut count = count;
    while count > 0 {
        res = cs[j] as i32 - ct[j] as i32;
        if res != 0 {
            break;
        }
        j += 1;
        count -= 1;
    }
    res
}

fn memscan(addr: &[u8], c: u8) -> usize {
    let mut i = 0;
    while i < addr.len() {
        if addr[i] == c {
            return i;
        }
        i += 1;
    }
    i
}

fn strstr_(s1: &[u8], s2: &[u8]) -> isize {
    let l2 = s2.len();
    if l2 == 0 {
        return 0;
    }
    let mut l1 = s1.len();
    let mut i = 0isize;
    while l1 >= l2 {
        l1 -= 1;
        if memcmp_(&s1[i as usize..i as usize + l2], s2, l2) == 0 {
            return i;
        }
        i += 1;
    }
    -1
}

fn strnstr_(s1: &[u8], s2: &[u8], mut len: usize) -> isize {
    let l2 = s2.len();
    if l2 == 0 {
        return 0;
    }
    let mut i = 0isize;
    while len >= l2 {
        len -= 1;
        if memcmp_(&s1[i as usize..i as usize + l2], s2, l2) == 0 {
            return i;
        }
        i += 1;
    }
    -1
}

fn memchr_(s: &[u8], c: u8, mut n: usize) -> isize {
    let mut i = 0isize;
    while n != 0 {
        n -= 1;
        if c == s[i as usize] {
            return i;
        }
        i += 1;
    }
    -1
}

fn check_bytes8(start: &[u8], value: u8, bytes: u32) -> isize {
    for i in 0..bytes as usize {
        if start[i] != value {
            return i as isize;
        }
    }
    -1
}

fn memchr_inv(start: &[u8], c: u8, bytes: usize) -> isize {
    let value = c;

    if bytes <= 16 {
        return check_bytes8(start, value, bytes as u32);
    }

    let mut value64: u64 = value as u64;
    value64 = value64.wrapping_mul(0x0101010101010101);

    let mut off = 0usize;
    let mut bytes = bytes;

    let mut prefix = (start.as_ptr() as usize) % 8;
    if prefix != 0 {
        prefix = 8 - prefix;
        let r = check_bytes8(&start[off..], value, prefix as u32);
        if r >= 0 {
            return off as isize + r;
        }
        off += prefix;
        bytes -= prefix;
    }

    let mut words = bytes / 8;
    while words != 0 {
        let w = u64::from_ne_bytes(start[off..off + 8].try_into().unwrap());
        if w != value64 {
            let r = check_bytes8(&start[off..], value, 8);
            return off as isize + r;
        }
        off += 8;
        words -= 1;
    }

    let r = check_bytes8(&start[off..], value, (bytes % 8) as u32);
    if r >= 0 {
        off as isize + r
    } else {
        -1
    }
}

// Identical LCG to diff_string.c.
struct Lcg(u64);
impl Lcg {
    fn next(&mut self) -> u32 {
        self.0 = self
            .0
            .wrapping_mul(6364136223846793005)
            .wrapping_add(1442695040888963407);
        (self.0 >> 32) as u32
    }
}

const ALPHABET: &[u8] = b"abcXYZ .";

fn gen_str(rng: &mut Lcg, maxlen: usize) -> Vec<u8> {
    let len = 1 + (rng.next() as usize) % (maxlen - 1);
    let mut v = Vec::with_capacity(len + 1);
    for _ in 0..len {
        v.push(ALPHABET[(rng.next() as usize) % ALPHABET.len()]);
    }
    v.push(0);
    v
}

fn gen_bytes(rng: &mut Lcg, len: usize, biased: bool) -> Vec<u8> {
    (0..len)
        .map(|_| if biased { (rng.next() % 3) as u8 } else { (rng.next() & 0xff) as u8 })
        .collect()
}

const BUFLEN: usize = 64;

fn cstr(v: &[u8]) -> &str {
    let end = v.iter().position(|&b| b == 0).unwrap_or(v.len());
    std::str::from_utf8(&v[..end]).unwrap()
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let n: u64 = args.get(1).and_then(|a| a.parse().ok()).unwrap_or(5000);
    let seed: u64 = args.get(2).and_then(|a| a.parse().ok()).unwrap_or(271828);
    let mut rng = Lcg(seed);

    for _ in 0..n {
        let a = gen_str(&mut rng, BUFLEN - 1);
        let b = gen_str(&mut rng, BUFLEN - 1);
        let len = (rng.next() as usize) % (BUFLEN - 1) + 1;

        println!(
            "strncasecmp,{},{},{},{}",
            cstr(&a),
            cstr(&b),
            len,
            strncasecmp(&a, &b, len)
        );
        println!("strcasecmp,{},{},{}", cstr(&a), cstr(&b), strcasecmp(&a, &b));

        let mut dst1 = vec![0u8; BUFLEN];
        let mut a_padded = a.clone();
        a_padded.resize(BUFLEN, 0);
        strncpy_(&mut dst1, &a_padded, BUFLEN - 1);
        println!("strncpy,{},[{}]", cstr(&a), cstr(&dst1[..BUFLEN - 1]));

        let mut dst2 = vec![0x55u8; BUFLEN];
        let cnt = (rng.next() as usize) % (BUFLEN - 1) + 1;
        let mut a_scpy = a.clone();
        a_scpy.resize(BUFLEN, 0);
        let r = sized_strscpy(&mut dst2, &a_scpy, cnt);
        println!("strscpy,{},{},{},[{}]", cstr(&a), cnt, r, cstr(&dst2));

        let mut dst3: Vec<u8> = a[..a.len() - 1].to_vec();
        dst3.push(0);
        strncat_(&mut dst3, &b, (rng.next() % 16) as usize);
        println!("strncat,{},{},[{}]", cstr(&a), cstr(&b), cstr(&dst3));

        let mut dst4: Vec<u8> = a[..a.len() - 1].to_vec();
        dst4.push(0);
        let src_nonul = &b[..b.len() - 1];
        let lr = strlcat_(&mut dst4, src_nonul, BUFLEN);
        println!("strlcat,{},{},{},[{}]", cstr(&a), cstr(&b), lr, cstr(&dst4));

        let ch = ALPHABET[(rng.next() as usize) % ALPHABET.len()];
        let r1 = strchrnul(&a, ch);
        println!("strchrnul,{},{},{}", cstr(&a), ch as char, r1);
        let r2 = strnchrnul(&a, len, ch);
        println!("strnchrnul,{},{},{},{}", cstr(&a), len, ch as char, r2);
        let r3 = strnchr(&a, len, ch);
        println!("strnchr,{},{},{},{}", cstr(&a), len, ch as char, r3);

        println!("strspn,{},{},{}", cstr(&a), cstr(&b), strspn(&a, &b));
        println!("strcspn,{},{},{}", cstr(&a), cstr(&b), strcspn(&a, &b));
        let r4 = strpbrk(&a, &b);
        println!("strpbrk,{},{},{}", cstr(&a), cstr(&b), r4);

        let mut sp: Option<&[u8]> = Some(&a[..]);
        print!("strsep,{},{}", cstr(&a), cstr(&b));
        for _ in 0..5 {
            if sp.is_none() {
                break;
            }
            let tok = strsep(&mut sp, &b[..]);
            print!(",[{}]", cstr(tok.unwrap()));
        }
        println!();

        let m1 = gen_bytes(&mut rng, BUFLEN, true);
        let mut m2 = m1.clone();
        let mlen = (rng.next() as usize) % BUFLEN;
        if rng.next() % 4 == 0 && mlen > 0 {
            let idx = (rng.next() as usize) % mlen;
            m2[idx] ^= 0xff;
        }
        println!("memcmp,{},{}", mlen, memcmp_(&m1[..mlen], &m2[..mlen], mlen));

        let sbuf = gen_bytes(&mut rng, BUFLEN, true);
        let target = (rng.next() % 3) as u8;
        let sr = memscan(&sbuf, target);
        println!("memscan,{},{}", target, sr);

        let sr2 = strstr_(&a[..a.len() - 1], &b[..b.len() - 1]);
        println!("strstr,{},{},{}", cstr(&a), cstr(&b), sr2);
        let slen = (rng.next() as usize) % (BUFLEN - 1);
        let mut a_ns = a.clone();
        a_ns.resize(BUFLEN, 0);
        let sr3 = strnstr_(&a_ns, &b[..b.len() - 1], slen);
        println!("strnstr,{},{},{},{}", cstr(&a), cstr(&b), slen, sr3);

        let cr = memchr_(&m1, target, mlen);
        println!("memchr,{},{},{}", target, mlen, cr);

        let fillval = (rng.next() % 3) as u8;
        let mut invbuf = vec![fillval; BUFLEN];
        if rng.next() % 3 != 0 {
            let pos = (rng.next() as usize) % BUFLEN;
            invbuf[pos] = (fillval + 1) % 3;
        }
        let ilen = (rng.next() as usize) % BUFLEN;
        let ir = memchr_inv(&invbuf[..ilen], fillval, ilen);
        println!("memchr_inv,{},{},{}", fillval, ilen, ir);
    }
}
